MNT: make Axes.cla an alias for Axes.clear in all cases

lijake8 opened the pull request:
## PR Summary
We (myself and @neil-gurnani) consolidated all implementations of cla/clear under a clear method (the clear functionality is implemented in clear, and cla calls clear for the sake of minimizing code duplication and improving maintainability). We added 2 tests ensuring subclasses do not override cla. Existing tests for cla were left as-is; since cla calls clear, coverage is not compromised.

Every subclass has both cla and clear, since it was mentioned in #22738 that cla is still too widely used to be fully deprecated as of now. We added appropriate docstrings/comments for all cla/clear methods.

We left all original '_api.deprecated' decorators as-is but did not add any new decorators (please let us know if this is needed).

Finally, please let us know if there's any issue with this PR we should address. This was our first PR and we really enjoyed working on it. 

Thanks

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

greglucas commented on examples/misc/custom_projection.py:
```diff
@@ -71,6 +72,10 @@ def cla(self):
         Axes.set_xlim(self, -np.pi, np.pi)
         Axes.set_ylim(self, -np.pi / 2.0, np.pi / 2.0)
 
+    def cla(self):
+        """Clear the current axes."""
+        self.clear()
+
```
Examples probably don't need to define cla() at all.

greglucas commented on lib/matplotlib/projections/geo.py:
```diff
@@ -55,6 +56,10 @@ def cla(self):
         Axes.set_xlim(self, -np.pi, np.pi)
         Axes.set_ylim(self, -np.pi / 2.0, np.pi / 2.0)
 
+    def cla(self):
+        """Clear the current axes."""
+        self.clear()
```
I think the `cla()` here can be inherited from the baseclass? So make sure that `Axes` has defined `cla()` to call `self.clear()`, which will then call your redefined `self.clear()` above. I'm pretty sure you already have this, so removing this method should be fine.

greglucas commented on lib/matplotlib/projections/geo.py:
```diff
@@ -484,12 +489,17 @@ def __init__(self, *args, center_longitude=0, center_latitude=0, **kwargs):
+        super().clear()
         self.yaxis.set_major_formatter(NullFormatter())
 
+    def cla(self):
+        """Clear the current axes."""
+        self.clear()
```
This can be removed too.

greglucas commented on lib/matplotlib/projections/polar.py:
```diff
@@ -373,12 +373,14 @@ def _wrap_locator_formatter(self):
         self.set_ticks_position('none')
         self._wrap_locator_formatter()
 
     @_api.deprecated("3.4", alternative="ThetaAxis.clear()")
     def cla(self):
+        """Clear the current axes."""
         self.clear()
```
Since this was deprecated in 3.4, it can be removed in 3.6 which is our next release. So, you can remove this here if you want. You'd need to add a doc api_change_note for removals indicating that it has been removed too.

greglucas commented on lib/matplotlib/projections/polar.py:
```diff
@@ -671,12 +673,14 @@ def _wrap_locator_formatter(self):
         self.set_ticks_position('none')
         self._wrap_locator_formatter()
 
     @_api.deprecated("3.4", alternative="RadialAxis.clear()")
     def cla(self):
+        """Clear the current axes."""
         self.clear()
```
Remove this as well if you want along with an API-change note.

greglucas commented on lib/matplotlib/projections/polar.py:
```diff
@@ -800,6 +805,10 @@ def cla(self):
         self.set_theta_offset(self._default_theta_offset)
         self.set_theta_direction(self._default_theta_direction)
 
+    def cla(self):
+        """Clear the current axes."""
+        self.clear()
```
Can remove this definition

greglucas commented on lib/matplotlib/spines.py:
```diff
@@ -224,6 +224,7 @@ def clear(self):
 
     @_api.deprecated("3.4", alternative="`.Spine.clear`")
     def cla(self):
+        """Clear the Axes."""
         self.clear()
```
Remove with an API change removal note if you want as well.

greglucas commented on lib/mpl_toolkits/axes_grid1/axes_grid.py:
```diff
@@ -37,11 +37,17 @@ def toggle_label(self, b):
+        super().clear()
         self.orientation = orientation
 
+    def cla(self):
+        """Clear the current axes."""
+        self.clear()
```
I think this instance of `cla()` does need to be kept around because this class isn't inheriting from Axes.

greglucas commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -468,6 +469,12 @@ def test_inverted_cla():
 
 
+def test_cla_not_redefined():
+    for klass in Axis.__subclasses__():
+        # check that subclasses do not get redefined in our Axis subclasses
```
```suggestion
    for klass in Axes.__subclasses__():
        # check that subclasses do not get redefined in our Axes subclasses
```
I think you want the "e" Axes instead of Axis?

greglucas reviewed (commented):
Looks like a great start! I think there were a few places that you added in a definition of `cla()` into a subclass when it could be inherited from the superclass, so I'd suggest looking at what each of the classes is inheriting from and making sure that there is a cla() in the parent that can get used. Let me know if that doesn't make sense or if I'm mistaken too!

Also, if you want to add your friend as a contributor on the commit, you can use the "co-authored-by" feature in the commit message: https://docs.github.com/en/github-ae@latest/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

lijake8 commented:
> Looks like a great start! I think there were a few places that you added in a definition of `cla()` into a subclass when it could be inherited from the superclass, so I'd suggest looking at what each of the classes is inheriting from and making sure that there is a cla() in the parent that can get used. Let me know if that doesn't make sense or if I'm mistaken too!

@greglucas Thanks for the feedback and suggestions! I just amended the PR. Summary:
- The cla() methods are now removed from all axes subclasses. 
- The test case is also updated and (locally) is verified to be passing. 
- Relevant deprecations have been removed and logged in the doc/api/next_api_changes/removals folder. 

Please let me know if there's anything else I should do!

greglucas commented on lib/mpl_toolkits/axes_grid1/parasite_axes.py:
```diff
@@ -17,11 +17,16 @@ def __init__(self, parent_axes, aux_transform=None,
         martist.setp(self.get_children(), visible=False)
         self._get_lines = self._parent_axes._get_lines
 
+    def cla(self):
+        """Clear the current axes."""
+        self.clear()
```
```suggestion
```
This is an interesting one because it looks like this isn't inheriting from Axes, but down below there is a class factory to create the ParasiteAxes from Axes, which I think means this does actually get a `cla()` from Axes when used in practice.

neil-gurnani replied:
Removing the `def cla(self)` because `ParasiteAxes` inherits from Axes with the class factory, as seen on lines 97-99 of `lib/mpl_toolkits/axes_grid1/parasite_axes.py`.

greglucas commented on lib/mpl_toolkits/axes_grid1/parasite_axes.py:
```diff
@@ -138,10 +143,15 @@ def draw(self, renderer):
+            ax.clear()
+        super().clear()
+
+    def cla(self):
+        """Clear the current axes."""
+        self.clear()
```
```suggestion
```

neil-gurnani replied:
Removing the `def cla(self)` because `HostAxes` also inherits from Axes with the class factory, as seen on lines 239-242 of `lib/mpl_toolkits/axes_grid1/parasite_axes.py`. 
Note: line numbers specified are before removal of the function.

greglucas commented on lib/mpl_toolkits/axisartist/floating_axes.py:
```diff
@@ -342,6 +343,10 @@ def cla(self):
         self.patch.set_clip_path(orig_patch)
         self.gridlines.set_clip_path(orig_patch)
 
+    def cla(self):
+        """Clear the current axes."""
+        self.clear()
```
```suggestion
```

neil-gurnani replied:
Removing the `def cla(self)` because `FloatingAxes` also inherits from Axes with the class factory, as seen on line 359-363 of `lib/mpl_toolkits/axisartist/floating_axes.py`.

neil-gurnani commented on lib/matplotlib/pyplot.py:
```diff
@@ -1108,7 +1108,12 @@ def sca(ax):
+    clear()
+
+
+def clear():
+    """Clear the current axes."""
+    return gca().clear()
```
Resolved, updated the `clear()` function in `pyplot.py` to use gca().cla().

greglucas replied:
I would just not add the `clear()` function to pyplot. So, don't touch this file at all in your changes in the PR.

It looks like your updates didn't get pushed up to the remote, I only see a new merge commit after your comments.

neil-gurnani commented on lib/matplotlib/tests/test_figure.py:
```diff
@@ -791,12 +791,18 @@ def test_figure_clear(clear_meth):
         assert 'clf' not in klass.__dict__
 
 
+def test_cla_not_redefined():
+    for klass in FigureBase.__subclasses__():
+        # check that subclasses do not get redefined in our Figure subclasses
+        assert 'cla' not in klass.__dict__
```
Updated `lib/matplotlib/tests/test_figure.py` to remove the `test_cla_not_redefined()` for the reason specified.

lijake8 commented:
Just pushed our updates. Unfortunately, we accidentally fetched from upstream before pushing and as a result there are some changes (particularly in pyplot) that are from upstream (the root matplotlib repo). Hopefully this is ok - if not, is there any way around this?

neil-gurnani commented:
@greglucas Thanks for the feedback and suggestions! When trying to merge my changes to the current commit, I kept running into Failed to Push Refs due to stale info errors as shown [here](https://stackoverflow.com/questions/56191415/why-is-git-push-force-with-lease-failing-with-rejected-stale-info-even). Then I tried to "fetch the updates from upstream and rebase first" as outlined in [this source](fetch the updates from upstream and rebase first), which caused the merge from main commit you saw. 

Anyways, my partner (@lijake8) was able to amend our changes for this PR. 

Summary:

- Reverted `pyplot.py` file to the pre-PR version with no addition of a `clear()` function

- Removed the `test_cla_not_redefined()` function from the `test_figure.py` file.

- Removed the def cla(self) for `ParasiteAxes` , `HostAxes`, and `FloatingAxes` as these three inherit from the `Axes` parent class using the`ClassFactory` method. 

Please let me know if there's anything else I should do!

lijake8 commented:
FInally fixed the commits issue, and the relevant tests to the changes pass, but now the automated main tests aren't passing because
`lib/matplotlib/tests/test_pyplot.py::test_pyplot_up_to_date`
fails. What should we do? This will resolve after merging with the main branch, right?

greglucas commented on lib/matplotlib/tests/test_figure.py:
```diff
@@ -796,7 +796,6 @@ def test_clf_not_refedined():
         # check that subclasses do not get redefined in our Figure subclasses
         assert 'clf' not in klass.__dict__
 
-
```
Hm... Is flake8 not getting run on tests? I'm pretty sure this line should be there. This file should also be left alone if you aren't touching it anymore.

lijake8 replied:
Fixed

greglucas commented on lib/mpl_toolkits/axisartist/axislines.py:
```diff
@@ -499,12 +499,13 @@ def _init_gridlines(self, grid_helper=None):
-        # gridlines need to b created before cla() since cla calls grid()
+    def clear(self):
+        # docstring inherited
+        # gridlines need to b created before clear() since clear calls grid()
```
```suggestion
        # gridlines need to be created before clear() since clear calls grid()
```
Might as well fix the typo while we're at it.

lijake8 replied:
Fixed

greglucas reviewed (approved):
Looks good to me! Just two very minor comments here. Then we just need to wait for a second reviewer before merging.

lijake8 commented:
@timhoffm @oscargus Do either of you mind reviewing this PR? Thanks!

timhoffm commented:
Thanks @lijake8 and congratulations on your first contribution to Matplotlib! We hope to see you again.

neil-gurnani commented:
@timhoffm My partner @lijake8 and I both worked on contributing to this issue together. 

He also added my name as a co-author as instructed by @greglucas above: 
"Also, if you want to add your friend as a contributor on the commit, you can use the "co-authored-by" feature in the commit message: https://docs.github.com/en/github-ae@latest/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors"

Can you confirm that both myself @neil-gurnani (Neil Gurnani) and my partner (@lijake8) will be credited for this contribution?

Thank you for your guidance with this and your approval of our Pull Request!

timhoffm commented:
> Can you confirm that both myself @neil-gurnani (Neil Gurnani) and my partner (@lijake8) will be credited for this contribution?

You’ve not followed the format that GitHub defines for co-author attribution. So on the technical side your account is not linked to the PR. OTOH your name is in the commit message, which states that you have contributed.

tacaswell commented:
As Tim said, it looks like the commit message did not trigger the co-author logic in GH.  I suspect the issue is that GH is looking for the tags at the end, but you put it in the first line of the commit message (that normally says what it is you did).

On one hand I am sympathetic to wanting to get this fixed and get your name listed on GH as a contributor.  On the other hand I do not think we want to do a force-push to main to re-write history to fix the commit message (we _have_ force pushed to main in the past but it is my judgement that this does not cross the threshold to do that) and I do not think that adding an empty commit with your name on it is a good option either (it makes it really odd if anyone looks it up later and I think is a bad precedent to set). 

Given that the concept of co-authoring is a GH concept (rather than a pure git concept), it may be worth reaching out to github support about this.   At a minimum they can add a warning to their docs about where the line is searched for.
